include/turtle/detail/mutex.hpp

```diff
@@ -36,25 +38,33 @@ namespace detail
 
     struct lock
     {
+    private:
+        BOOST_MOVABLE_BUT_NOT_COPYABLE(lock)
+
+    public:
         lock( const boost::shared_ptr< mutex >& m )
             : m_( m )
         {
             m_->lock();
         }
-        lock( const lock& rhs )
-        {
-            m_.swap( rhs.m_ );
-        }
         ~lock()
         {
             if( m_ )
                 m_->unlock();
         }
-
+        lock( BOOST_RV_REF( lock ) x )
+            : m_( x.m_ )
+        {
+            x.m_.reset();
```

Comment:
Shouldn't that be a `boost::move`?
```
    m_( boost::move( x.m_ ) )
```
?

Reply:
No. I need deterministic behaviour here: The other value **must** be NULL so the lock is not unlocked twice. I had the move there and found at least 1 implementation of shared_ptr that did not reset the moved-from value to NULL

Reply:
Sounds to me like a bug in that implementation of shared_ptr, but fair enough.

Reply:
Could be a swap here in the constructor I suppose, but never mind :)

Reply:
> Sounds to me like a bug in that implementation of shared_ptr, but fair enough.

It isn't. A "moved from value" has undefined, but valid state. So in the case of a shared_ptr it could do a simple copy and be correct. If at all it is a missed optimization possibility. Still: You can't rely on it being a specific value, so better be safe. Reset is used in both cases for consistency.